### DISCUSSION TITLE
getting latest memtest files from web site and setting up for grub2?

### DISCUSSION THREAD

**Discussion by msetzerii** [memtest86plus/memtest86plus#314] [General]
Redhat/Fedora has a discussion on getting 6.20 added as rpm, but is complex discussion.

I've done a manual process that extracts the two memtest programs from iso file. With grub4dos I've found that legacy boot needs the memtest from boot, while uefi option needs.
The grub2 seems to work fine with either of the memtest programs.

I'm trying to get a script that will automate process and support newer versions and extraction of the programs.

Crude version so far getmemtest.sh

cd /tmp
mkdir memtest
cd memtest
rm index.html -f
wget https://memtest.org/ 
f=$(grep download <index.html | grep grub.iso | sed 's/[ \t]//g'|cut -f2 -d\" |head -n 1)
rm index.html -f
wget -nc https://memtest.org/$f
echo $f
zip=$(echo $f | cut -f4 -d/)
unzip -u $zip
iso=$(echo $zip | sed 's/.zip//g')
boot=$(find / -maxdepth 2 -iname loader | sed 's_/loader__')
mount -o loop $iso /mnt
cp /mnt/boot/memtest $boot/memtest.leg -p -v
cp /mnt/EFI/BOOT/memtest $boot/memtest.uefi -p -v
umount /mnt

Downloads the index.html file from site to get the latest path to version. Currently 6.20
Then it downloads the zip file.
extracts zip file to have iso file.
Then mounts the iso file 
copies each of the memtest files to memtest.leg an memtest.uefi in boot directory.
Seems there is some setups that don't have /boot as a separate partition??

I've replaced the original 5.31 memtest setup file in /etc/grub.d/20_memtest86+

# !/bin/sh
exec tail -n +3 $0
# This file provides an easy way to add custom menu entries.  Simply type the
# menu entries you want to add after this comment.  Be careful not to change
# the 'exec tail' line above.
menuentry 'Memtest Legacy' {
	echo 'Loading memtest 6.20 for BIOS BOOT SYSTEM'
	linux /memtest.leg
}
menuentry 'Memtest UEFI' {
	echo 'Loading memtest 6.20 for UEFI BOOT SYSTEM'
	linux /memtest.uefi
}

Currently has hardcode 6.20, so could just remove that, or work to modify it with new version number.
Have used 6.01, 6.10, and 6.20.

But at least there is discusion on it. The old 5.31 version that have hasn't worked for long time, and for UEFI the setup just says it isn't support. The 6.x versions work with both legacy and uefi boot (secure boot turned off).

Just wondering if someone else might have better solution or recommenations.
Thanks.

Use the memtest86+ on my g4l project since memory is critical for disk imaging. So having a way to make sure ram is good i nice. Thanks for the great program.

**Comment by debrouxl** [comment]
From a quick glance, if I wanted to extract memtest86+ release binaries from the official site instead of:
* either using the contents of my distro's packages, which are WIP for Fedora, per your message;
* or recompiling memtest86+ by myself, which would give me the ability to cherry-pick some relevant bugfixes made since the latest release, or apply some of my own size optimization patches

then I'd do something along those lines as well, possibly with a temporary directory created using $(mktemp) instead of a temporary directory with a fixed and therefore predictable name.

**Comment by x86fr** [comment]
You don't have to extract from ISO, Memtest86+ binaries (UEFI and BIOS) are available in this archive: https://memtest.org/download/v6.20/mt86plus_6.20.binaries.zip

**Reply by msetzerii** [reply]
Modified the script to use the binary option and have it get the latest version from file to create a 20_memtest86+ file that shows version. Still requires running grub2-mkconfig to actually create the updated grub.cfg 

\#!/usr/bin/bash
cd /tmp
mkdir memtest
cd memtest
rm index.html -f
wget https://memtest.org/ 
f=$(grep download <index.html | grep binaries.zip | sed 's/[ \t]//g'|cut -f2 -d\" |head -n 1)
rm index.html -f
wget -nc https://memtest.org/$f
echo $f
zip=$(echo $f | cut -f4 -d/)
unzip -u $zip
boot=$(find / -maxdepth 2 -iname loader | sed 's_/loader__')
cp memtest64.bin $boot/memtest.leg -p -v
cp memtest64.efi $boot/memtest.uefi -p -v
ver=$(echo $zip | cut -b10-13);
{
printf "# \x21/bin/sh\nexec tail -n +3 \$0\n# This file provides an easy way to add custom menu entries.  Simply type the\n";
printf "# menu entries you want to add after this comment.  Be careful not to change\n# the 'exec tail' line above.\n" ;
printf "menuentry 'Memtest Legacy' {\n	echo 'Loading memtest $ver for BIOS BOOT SYSTEM'\n	linux /memtest.leg\n}" ;
printf "\nmenuentry 'Memtest UEFI' {\n	echo 'Loading memtest $ver for UEFI BOOT SYSTEM'\n	linux /memtest.uefi\n}\n";  } >/etc/grub.d/20_memtest86+
chmod 755 /etc/grub.d/20_memtest86+

Have been some message under bugzilla for redhat on creating a 6.20 rpm, but it has some issues trying to handle all setups..
Have this work on 5 of my Fedora systems. Works more like the 5.31 rpm that installs files and then uses memtest-setup to actually create the 20_memtest86+ file and old setup had memtest-setup.

The messages on bugzilla list a 5/19 version verse the 5/7 date for files from web site. Talks about some patches??
The 6.20 also has some other files it puts under /usr/share
./doc
./doc/memtest86+
./doc/memtest86+/README.md
./licenses
./licenses/memtest86+
./licenses/memtest86+/LICENSE

**Comment by msetzerii** [comment]
Thanks. Had used iso before and didn't know exactly 
what th binaries.zip had.

Modified script to handle it.
getmemtest2.sh

#!/usr/bin/bash
cd /tmp
mkdir memtest
cd memtest
rm index.html -f
wget https://memtest.org/ 
f=$(grep download <index.html | grep binaries.zip | sed 's/[ \t]//g'|cut -f2 -d\" |head -n 1)
rm index.html -f
wget -nc https://memtest.org/$f
echo $f
zip=$(echo $f | cut -f4 -d/)
unzip -u $zip
boot=$(find / -maxdepth 2 -iname loader | sed 's_/loader__')
cp memtest64.bin $boot/memtest.leg -p -v
cp memtest64.efi $boot/memtest.uefi -p -v

**Reply by x86fr** [reply]
I would recommend using the 32-bit binary for legacy BIOS (more universal for older PCs)

**Comment by msetzerii** [comment]
On 22 May 2023 at 10:10, Sam Demeulemeester wrote:

Date sent:	Mon, 22 May 2023 10:10:36 -0700
From:	Sam Demeulemeester ***@***.***>
Send reply to:	memtest86plus/memtest86plus ***@***.***>
To:	memtest86plus/memtest86plus ***@***.***>
Copies to:	msetzerii ***@***.***>, 
	Author ***@***.***>
Subject:	Re: [memtest86plus/memtest86plus] getting latest memtest files from
	web site and setting up for grub2? (Discussion #314)

> 
> 
> I would recommend using the 32-bit binary for legacy BIOS (more universal 
> for older PCs)

Thanks for reply. Will have to look at that. Had be using 
32bit version of Fedora to build my g4l project going 
back to 2004, and kept it that way to support both 32 
and 64 bit. Did a test once, and saw no real addition for 
64bit version for disk imaging. But then they dropped 
32bit support, so had to switch everything to 64bit 
serveral versions ago. 

My project has syslinux boot version, and a verion that 
has grub4dos with both legacy and uefi loaders. Think it 
is 64 bit version, so 32 bit machines wouldn't work with 
it anyway. 

Thanks again.


> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.


+------------------------------------------------------------+
 Michael D. Setzer II - Computer Science Instructor 
(Retired)     
 ***@***.***                            
 ***@***.***
 Guam - Where America's Day Begins                        
 G4L Disk Imaging Project maintainer 
 http://sourceforge.net/projects/g4l/
+------------------------------------------------------------+
